File: q151-175/q153.py
# 2021.04.02
# Problem Statement:
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

class Solution:
    def findMin(self, nums: List[int]) -> int:
        # early check for single-element or already sorted lists
        if len(nums) == 1 or nums[0] < nums[-1]: return nums[0]
        
        # binary search
        left = 0
        right = len(nums)
        
        while True:
            middle = int(1/2 * (left + right))
            # do not need to consider when the middle is 0, cause it's already checked in the beginning
            if middle == len(nums)-1 or (nums[middle]<nums[middle-1] and nums[middle]<nums[middle+1]):
                return nums[middle]
            else:
                if nums[middle]>nums[0]:
                    # search the right half
                    left = middle
                else:
                    # search the left half
                    right = middle
        
        
        
                
                

File: Blind75/q153.java
// 2022.09.10
// Problem Statement:
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

// idea: binary search, if not the min, compare with the last element in the array,
// if larger, search right half; if smaller, search the left half
class Solution {
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length-1;
        while (left<=right) {
            if (left==right) return nums[left];
            int middle = left + (right-left)/2;
            if (middle==0 && nums[middle]<=nums[nums.length-1]) return nums[middle];
            else if (middle>0 && nums[middle]<nums[middle-1]) return nums[middle];
            else if (nums[middle]>nums[right]) { // search right half
                left = middle+1;
            } else { // search the left half
                right = middle-1;
            }
        }
        return 0;
    }
}